src/machine/aot/mod.rs

```diff
@@ -0,0 +1,673 @@
+        Ok(())
+    }
+
+    fn store64(&mut self, _addr: &Value, _value: &Value) -> Result<(), Error> {
+        Ok(())
+    }
+}
+
+pub struct AotCode {
+    pub code: Mmap,
+    /// Labels that map RISC-V addresses to offsets into the compiled x86_64
+    /// assembly code. This can be used as entrypoints to start executing in
+    /// AOT code.
+    pub labels: HashMap<u64, u32>,
+}
+
+impl AotCode {
+    pub fn base_address(&self) -> u64 {
+        self.code.as_ptr() as u64
+    }
+}
+
+pub struct AotCompilingMachine {
+    registers: [Value; 32],
+    pc: Value,
+    emitter: Emitter,
+    memory: FlatMemory<u64>,
+    sections: Vec<(u64, u64)>,
+    dummy_sections: HashMap<u64, u64>,
+    addresses_to_labels: HashMap<u64, u32>,
+    writes: Vec<Write>,
+    next_pc_write: Option<Value>,
+    instruction_cycle_func: Option<Box<InstructionCycleFunc>>,
+}
+
+impl AotCompilingMachine {
+    pub fn load(
+        program: &Bytes,
+        instruction_cycle_func: Option<Box<InstructionCycleFunc>>,
+    ) -> Result<Self, Error> {
+        // First we need to gather labels
+        let mut label_gathering_machine = LabelGatheringMachine::load(&program)?;
+        label_gathering_machine.gather()?;
+
+        let mut labels: Vec<u64> = label_gathering_machine.labels.iter().cloned().collect();
+        labels.sort_unstable();
+        let addresses_to_labels = labels
+            .iter()
+            .enumerate()
+            .map(|(i, address)| (*address, i as u32))
+            .collect();
+
+        Ok(Self {
+            registers: init_registers(),
+            pc: Value::from_u64(0),
+            emitter: Emitter::new(labels.len())?,
+            addresses_to_labels,
+            memory: label_gathering_machine.memory,
+            sections: label_gathering_machine.sections,
+            dummy_sections: label_gathering_machine.dummy_sections,
+            writes: vec![],
+            next_pc_write: None,
+            instruction_cycle_func,
+        })
+    }
+
+    fn read_pc(&self) -> Result<u64, Error> {
+        match &self.pc {
+            Value::Imm(pc) => Ok(*pc),
+            _ => Err(Error::Unexpected),
+        }
+    }
+
+    fn emit_block(&mut self, instructions: &[Instruction]) -> Result<(), Error> {
+        let mut cycles = 0;
+        for i in instructions {
+            let pc = self.read_pc()?;
+            let length = instruction_length(*i);
+            cycles += self
+                .instruction_cycle_func
+                .as_ref()
+                .map(|f| f(*i))
+                .unwrap_or(0);
+            execute(*i, self)?;
+            self.pc = Value::from_u64(pc + u64::from(length));
+        }
+        let pc = self.read_pc()?;
+        self.emitter.emit_add_cycles(cycles)?;
+        // Emit succeeding PC write only
+        if pc >= RISCV_MAX_MEMORY as u64 {
+            return Err(Error::OutOfBound);
+        }
+        self.emitter.emit(&Write::Pc {
```

Comment:
Is it safe to move this to the else branch of `if let Some(value) = self.next_pc_write.take() {`.

Reply:
I thought about it but it will make the generated assembly hard to write. Hence I'm sticking to the simple way here. In fact I think if we really want to optimize this, we can fix the VM so we can skip this. But that will be a bigger change and requires changing our internal bytecode, so I want to leave it till another time.

Another point is this is the same behavior as our assembly interpreter.